### PR TITLE
Testing and reports

### PR DESCRIPTION
**Pull request by AA-Turner**
Couple of things in this one

- Move testing to parameterised using [hypothesis](https://github.com/HypothesisWorks/hypothesis). Shouldn't be any noticable difference, but this will be helpful for identifying edge cases etc as we move forwards (and increase test coverage!)

- Add functionality relating to #51. I'm still not happy with this as (1) it requires regenerating both the merged census extract and the reduced ONS file, and (2) it doesn't preserve population sizes etc. (2) is an easier fix, for (1) the ideal is to dynamically hook in without any pre-setup etc.

WIP for now as not ready for merging

### CODE REVIEW

File: src/data/ons_pd_may_18.py [hunk 1]
```diff
@@ -50,13 +49,11 @@ class ONSPostcodeDirectoryMay18(ONSPostcodeDirectory):
     }
 
     def __init__(self, ons_pd_csv_path, load_data=True):
-        ONSPostcodeDirectory.__init__(
-            self, ons_pd_csv_path, load_data, ONSPostcodeDirectoryMay18.index_column, ONSPostcodeDirectoryMay18.fields, ONSPostcodeDirectoryMay18.data_types,
-        )
+        super().__init__(ons_pd_csv_path, load_data, ONSPostcodeDirectoryMay18.index_column, ONSPostcodeDirectoryMay18.fields, ONSPostcodeDirectoryMay18.data_types)
 
         # Folder within the ONS Postcode Directory archive holding names and codes files
-        names_codes_root = DATA_ROOT / self.settings["ONS Names and codes folder"]
```
Comment (ox-baloo) [inline src/data/ons_pd_may_18.py:58]: did you mean to remove DATA_ROOT?

File: src/data/ons_pd_may_19.py [hunk 2]
```diff
@@ -50,13 +49,11 @@ class ONSPostcodeDirectoryMay19(ONSPostcodeDirectory):
     }
 
     def __init__(self, ons_pd_csv_path, load_data=True):
-        ONSPostcodeDirectory.__init__(
-            self, ons_pd_csv_path, load_data, ONSPostcodeDirectoryMay19.index_column, ONSPostcodeDirectoryMay19.fields, ONSPostcodeDirectoryMay19.data_types,
-        )
+        super().__init__(ons_pd_csv_path, load_data, self.index_column, ONSPostcodeDirectoryMay19.fields, ONSPostcodeDirectoryMay19.data_types)
 
         # Folder within the ONS Postcode Directory archive holding names and codes files
-        names_codes_root = DATA_ROOT / Path(self.settings["ONS Names and codes folder"]).resolve()
```
Comment (ox-baloo) [inline src/data/ons_pd_may_19.py:58]: did you mean to remove DATA_ROOT?

File: src/data/ons_pd_may_18.py (hunk 1 above)
Comment (AA-Turner) [reply src/data/ons_pd_may_18.py:58]: Yes I think so as `data/` is in `settings.json `already https://github.com/the-scouts/geo_mapping/blob/4c6484d041d7a1ebc0fca964d714b68504d80faf/settings.json#L5-L6
